### ISSUE TITLE
[improve-pipeline] Take one mechanism from SkillOpt: route the prior-verdict search ahead of the dialectic, make Reject runs durable

### ISSUE THREAD

**Issue by chrislacey89** [chrislacey89/skills#216]
**Target repo:** chrislacey89/skills
**Date:** 2026-08-12
**Status:** proposed
**Triggering repo:** none — reading session, not a downstream project
**Triggering context:** user request to evaluate arXiv 2605.23904v2 ("SkillOpt: Executive Strategy for Self-Evolving Agent Skills", Microsoft/SJTU, May 2026) for pipeline takeaways
**Pipeline area:** `/improve-pipeline` (meta-skill)
**Change type:** workflow
**Confidence:** medium

## Field Incident

Skill Kit improves itself through `/improve-pipeline`: a lesson comes in, three roles argue it out (Advocate, Skeptic, Mediator), and the result is filed as a GitHub issue for review. That loop is the repo's version of training — the skills are the weights, and the issues are the gradient steps.

The SkillOpt paper formalizes the same idea for benchmark skills: it treats a skill document as trainable external state of a frozen model, and optimizes it with deep-learning discipline. Most of its machinery needs things Skill Kit doesn't have (automatic verifiers, scored rollouts, held-out splits). **But one mechanism exposed a real wiring flaw in `/improve-pipeline` itself: SkillOpt feeds its rejected proposals back to the optimizer *before* it proposes again, while our equivalent — the related-issue search over 200+ prior verdicts — runs in Phase 5, *after* the Phase 4 dialectic has already argued and ruled.** The search can prevent a duplicate filing; it cannot inform the argument. And runs whose verdict is Reject have no clearly mandated durable trace at all, so a rejected evaluation costs a full run and leaves nothing the next run can find.

There is a live precedent for that second gap: the 16-framework audit (2026-08) left its rejected candidates in a gitignored `.context/` dossier in another workspace — negative results living outside the repo's own "state lives in GitHub" rule.

**Why it matters:** with a dense tracker, almost every new proposal area has prior verdicts. A dialectic that can't see them re-litigates settled questions at full cost, and a Skeptic without access to prior Rejects argues from a blank slate when the strongest evidence already exists.

## Pipeline Run Breakdown

- **Path run:** `/improve-pipeline` (this run: Phase 0 context load → paper analysis → Phase 3.5 library consult → Phase 4 sub-agent dialectic → this issue)
- **Expected behavior:** prior verdicts in the same area inform the dialectic before it argues
- **Actual behavior:** the `--state all` issue search runs only at filing time (Phase 5, `SKILL.md:227–237`), as a dedup check; Reject verdicts may leave no GitHub trace ("filing was intentionally deferred with a stated reason" is a sanctioned exit, `SKILL.md:404`)
- **Introduced at:** `/improve-pipeline`'s original phase ordering
- **Missed by:** every run since — the search always "works," so the misrouting is invisible
- **Cost:** repeated full-cost re-litigation of settled questions; rejected evaluations evaporating (or landing in gitignored dossiers)

## Why This Is a Pipeline Problem

The weakness is in the meta-skill's own structure, not in any downstream project. The information already exists (the issue tracker) and the query already exists (Phase 5's `gh issue list --search ... --state all`); only the routing is wrong. This is Meadows' leverage point #6 in miniature — the Dutch meter study: identical information, different routing, 30% behavior change. We are moving the meter from the basement to the hallway.

## Repo-Wide Context

- `improve-pipeline/SKILL.md` — the only file that changes; phase ordering (3.5 → 4 → 5), Advocate/Skeptic specs, Status template line, Verification list
- `docs/skill-anatomy.md` — the no-op test, sediment/sprawl vocabulary, and "per-sentence diagnostic, not a length target" — already encodes the paper's bounded-edit and compactness findings; constrains this change to a relocation, not an accretion
- `SYSTEM-OVERVIEW.md` § Handoff Table / `CLAUDE.md` § Invocation Roles — no changes; no skill is added, renamed, or re-roled
- `scripts/skill-references.manifest` — no change; no new `docs/`-rooted reference is bundled
- Adjacent: `/compound` (captures *project* lessons; untouched), `skill-creator` eval tooling (where any future measurement apparatus would belong — explicitly not imported here)

## Library Consultation

- **Books surveyed:** `~/.claude/library` index searched for systems/feedback/measurement handles
- **Books loaded:** `donella-meadows-thinking-in-systems` (leverage-points), `argyris-action-science` (core-framework), `goodhart-strathern` (core-mechanism)
- **Key principles applied:**
  - Meadows, leverage point #6: "Missing information flows is one of the most common causes of system malfunction. Adding or restoring information can be a powerful intervention, usually much easier and cheaper than rebuilding physical infrastructure." — the change is a routing fix, not new infrastructure
  - Argyris: "Organizations that continuously improve their processes within a fixed goal structure experience themselves as highly adaptive. They are not." — a dialectic blind to its own prior verdicts is single-loop learning; making prior decisions examinable inside the argument is the minimal double-loop upgrade
  - Goodhart/Strathern: "When a measure becomes a target, it ceases to be a good measure." — grounds the explicit rejection of SkillOpt's validation-gate and compactness machinery (pseudo-metrics on long-feedback-loop human outcomes)

## Suggested Further Reading

**From the library** (already available via `/library <name>`):

- `annie-duke-quit` — kill criteria and recording folds, not just bets played; the epistemics behind making Reject verdicts durable (Duke is already in this skill's `sources:`)
- `microsoft-skillopt` — **acquired; gap closed 2026-08-16.** 15 reference files. `results-and-ablations.md` is the ablation evidence this issue asked for. `core-framework.md` and `optimization-loop.md` for the mechanism being adopted; `validation-gate.md`, `anti-overfitting-layers.md`, and `two-timescale-updates.md` are the machinery this issue explicitly *rejects* — read them before re-litigating that rejection, since the Goodhart argument above was made without them.

**Gap in the library** — ~~not yet available; worth acquiring~~ **closed 2026-08-16**:

- ~~*SkillOpt: Executive Strategy for Self-Evolving Agent Skills* — Yang et al., arXiv 2605.23904 — belongs in `~/.claude/library/papers/`; it is the empirical companion to this repo's whole thesis (skills as trainable external state of a frozen agent), and its ablation tables are reusable evidence for future skill-editing debates~~ Now in the library as `microsoft-skillopt`. Installed under `books/`, not `papers/` — the location preference in the original bullet was not honored, but the content is present.

## Advocate Case

- The skill's only memory — 200+ prior verdicts — is routed to the wrong point in the run: Phase 5's search runs after the verdict, so it can only dedup, never inform. Prior Rejects/Defers are the one context class with no mandated load point before Phase 4 (Phase 0 loads docs, Phase 3 loads files, Phase 3.5 loads books).
- Reject runs are structurally lossy: the Verification list sanctions "filing intentionally deferred," so a full-cost dialectic can evaporate. The `.context/` dossier precedent shows negative results already leak out of GitHub.
- Issue #207 names the governing law: records without consumers get re-filed. Durable Rejects (B) *creates records*; search-before-dialectic (A) *creates their consumer*. They are one change, not two.
- The fix converts the Mediator's existing "what evidence would later justify revisiting" output from decoration into a live input — under the repo's own no-op test, the strongest kind of change: already-written prose starts changing behavior.
- Net line delta ≈ zero: one moved block, ~4 new sentences, ~6 removed — a routing change on a skill already flagged for bloat (#68), not an accretion.

## Skeptic Case

- The trigger fails the skill's own entry condition: no field incident, no observed instance of a dialectic re-proposing a rejected idea. The failure A/B prevents is hypothesized from a paper, not observed in the repo. Precedent: #68 was also a reading-driven proposal and sits open and unactioned.
- SkillOpt's buffer value (1.6–4.6 points) is in-distribution and same-task — thousands of rollouts against one verifier. `/improve-pipeline` proposals are heterogeneous one-offs; proposal density per topic is close to one.
- Precedent-creep risk: a Reject here is a contextual editorial judgment against a repo state that changes (#205/#206 invalidated a whole line of prior reasoning within two months). Required prior-verdict input risks treating stale verdicts as case law.
- Meadows, policy resistance: a mandated "consult prior verdicts" step degrades into a rote filler paragraph — #207's disease under a new heading. The moment A grows a required template slot, this objection reactivates.
- Everything else in the paper is duplicate or contradicted: bounded edits ≈ "prefer the smallest coherent fix" + the no-op test; strict-improvement gating without a verifier is Goodhart bait; compactness targets directly contradict skill-anatomy.md's "per-sentence diagnostic, not a length target."
- Steelman concession: A's marginal cost is near zero *as a relocation*; B's honest delta is one disambiguating sentence; if anything ships it is exactly that, in one issue.

## Mediator Verdict

**Verdict:** Proceed narrowly (candidates A+B as one routing change); Reject candidate C (importing anything else from the paper).

Both roles cited Meadows to opposite ends — leverage point #6 (add the missing information flow) for, policy resistance (mandated steps become filler) against. That tension is real and is resolved by the change's *shape*, not by picking a winner: A must be a **relocation** of the existing search block with **no new required template slot**, so there is no compliance surface for policy resistance to fill. The Skeptic's Defer-pending-an-incident position has a catch-22 the verdict must name: without durable Reject records and a pre-dialectic search, verdict re-litigation is unobservable, so the evidence a Defer waits for could never accumulate. B at minimum must land for the Defer condition to even be measurable; A is free once B exists.

Minimal coherent change set: single file, `improve-pipeline/SKILL.md`. What stays unchanged: no validation gates, no metrics, no compactness targets, no optimizer vocabulary, no changes to `/compound`, the dialectic itself, or any inventory surface. Falsification condition for revisiting: if post-change issues show the prior-verdict step producing rote "none relevant" lines rather than cited verdicts, the relocation failed to escape policy resistance and should be reverted to a filing-time check.

## Recommended Changes

| Target file | Change type | Change | Why | Risk if skipped |
|-------------|-------------|--------|-----|-----------------|
| `improve-pipeline/SKILL.md` | modify | New Phase 3.75 "Load prior verdicts" between Phase 3.5 (line 154) and Phase 4 (line 168): relocate the existing `gh issue list -R chrislacey89/skills --search "<keywords>" --state all` block from Phase 5; read the Mediator Verdict of the 1–3 closest hits; one sentence each in the Advocate and Skeptic specs requiring each role to address the closest prior verdict when one exists (prior verdicts are evidence to answer, not binding precedent) | Routes existing information to the decision point (Meadows #6); gives the Mediator's "revisit condition" output a consumer | Dialectic keeps arguing from a blank slate beside 200+ settled verdicts |
| `improve-pipeline/SKILL.md` | consolidate | Phase 5 retitled "File the GitHub Issue"; dedup survives as one line ("if a Phase 3.75 hit already covers the same problem, update or comment on it instead of filing; if related but distinct, cross-link") | Single source of truth — the search must not live in two phases | Two copies of the search rule drift silently (skill-anatomy.md, Sprawl) |
| `improve-pipeline/SKILL.md` | modify | Phase 5: a Reject verdict still files the issue (same full template — no lighter artifact to game) and immediately closes it as `not planned`; template `**Status:**` line (line 264) gains `rejected`; Verification bullet (line 404) tightened so a completed dialectic always leaves a GitHub trace | Negative results become searchable by the relocated Phase 3.75 query; removes the sanctioned-evaporation exit | Rejected evaluations keep leaking into gitignored dossiers or the maintainer's head |

Text removed (with location): from `### Phase 5: Search for Related Issues, Then File the GitHub Issue` (lines 227–237), relocated (not deleted) to Phase 3.75:

> Search for related issues in `chrislacey89/skills` before filing anything new.
>
> ```bash
> gh issue list -R chrislacey89/skills --search "<keywords>" --state all
> ```
>
> - Search open and closed issues using the pipeline area, skill names, and the clearest failure-mode keywords from the incident.
> - If an existing issue already covers the same problem, update or comment on it instead of filing a duplicate.
> - If an existing issue is related but distinct, file the new issue and cross-link both.

Text removed (with location): from `## Verification` (line 404), the clause "or filing was intentionally deferred with a stated reason" is narrowed — a completed dialectic always files; only runs stopped before Phase 4 (e.g., Phase 2 Local-event exits) may end without an issue.

## Non-Goals

- No validation gate, selection split, score, or any quantitative acceptance machinery (Goodhart)
- No token/length targets for SKILL.md files — skill-anatomy.md's per-sentence no-op test stands; the paper's ~920-token median is a benchmark artifact, and its bounded-edit finding is read as *directional validation* of existing doctrine, not as new text
- No optimizer vocabulary ("learning rate", "buffer", "gate") in skill prose — `/improve-pipeline` stays a deliberative editorial process, "not automatic self-healing"
- No changes to `/compound`, the dialectic structure (#64 remains separate), or any inventory surface

## Follow-On Options

- **Proposal only:** stop here and review the issue
- **If approved:** implement the three rows above in `chrislacey89/skills` as a single small PR

## Related Evidence

- arXiv 2605.23904v2 — SkillOpt; rejected-edit-buffer ablation −1.6 to −4.6 points; unbounded rewriting degrades vs bounded patches; skills transfer across harnesses (directional support for the pack's install-everywhere model)
- #207 — records without consumers get re-filed (the constraint this proposal designs around)
- #68 — `/improve-pipeline` bloat flag (why the change must be net-zero lines)
- #64 — dialectic sub-agent isolation (adjacent, distinct; this run used sub-agents for Advocate/Skeptic)
- #29 (closed) — precedent for meta-changes to this skill
- 2026-08 framework-audit dossier in a gitignored `.context/` workspace directory — field example of rejected evaluations escaping GitHub-native state

**Comment by chrislacey89** [comment]
## Library gap closed — `microsoft-skillopt`

The SkillOpt paper is in the library (15 reference files). The gap bullet is struck through, the book added to **From the library**, and the `library-gap` label removed.

**The specific ask is met.** The bullet wanted the ablation tables as "reusable evidence for future skill-editing debates" — that is `results-and-ablations.md`.

**One thing to flag, because it changes what this issue can now claim.** The rejection above — SkillOpt's validation-gate and compactness machinery as Goodhart-vulnerable pseudo-metrics on long-feedback-loop human outcomes — was written from the paper's abstract-level shape, not its contents. The library now has `validation-gate.md`, `anti-overfitting-layers.md`, and `two-timescale-updates.md`: the three rejected mechanisms, each with its own file. That rejection is now *checkable*, and it was not when the Mediator ruled. Anyone implementing should read those three before treating the rejection as settled — it may well survive, but it currently rests on an argument made without the source.

**Filed** 2026-08-12 14:53Z → **acquired** ~2026-08-12 20:02Z, about **5 hours later**. The issue was correct when written.

Installed under `library/books/`, not `library/papers/` as the bullet requested. Content is present; the location preference was not honored.
